lib/ipmi_strings.c

```diff
@@ -754,3 +769,251 @@ const struct oemvalstr picmg_busres_shmc_status_vals[] = {
 
  { 0xffffff, 0x00,  NULL }
 };
+
+
+/**
+ * A helper function to count repetitions of the same byte
+ * at the beginning of a string.
+ */
+static
+size_t count_bytes(const char *s, unsigned char c)
+{
+	size_t count;
+
+	for (count = 0; s && s[0] == c; ++s, ++count);
+
+	return count;
+}
+
+/**
+ * Parse the IANA PEN registry file.
+ *
+ * See https://www.iana.org/assignments/enterprise-numbers/enterprise-numbers
+ * The expected entry format is:
+ *
+ * Decimal
+ * | Organization
+ * | | Contact
+ * | | | Email
+ * | | | |
+ * 0
+ *   Reserved
+ *     Internet Assigned Numbers Authority
+ *       iana&iana.org
+ *
+ * That is, IANA PEN at position 0, enterprise name at position 2.
+ */
+#define IANA_NAME_OFFSET 2
+#define IANA_PEN_REGISTRY "enterprise-numbers"
+int ipmi_oem_info_init()
+{
+	FILE *in = NULL;
+	oem_valstr_list_t terminator = { { -1, NULL}, NULL }; /* Terminator */
+	oem_valstr_list_t *oemlist = &terminator;
+	/* Do not count terminators */
+	size_t head_entries = ARRAY_SIZE(ipmi_oem_info_head) - 1;
+	size_t tail_entries = ARRAY_SIZE(ipmi_oem_info_tail) - 1;
+	/* Include static entries and the terminator */
+	size_t count = head_entries + tail_entries + 1;
+	char *home;
+
+	lprintf(LOG_INFO, "Loading IANA PEN Registry...");
+
+	if (ipmi_oem_info) {
+		lprintf(LOG_INFO, "IANA PEN Registry is already loaded");
+		return 0;
+	}
+
+	/*
+	 * First try to open user's local registry if HOME is set
+	 */
+	if ((home = getenv("HOME"))) {
+		char path[PATH_MAX + 1] = { 0 };
+		snprintf(path, PATH_MAX, "%s%s",
+		         home,
+		         PATH_SEPARATOR IANAUSERDIR PATH_SEPARATOR IANA_PEN_REGISTRY);
+		in = fopen(path, "r");
+	}
+
+	if (!in) {
+		/*
+		 * Now open the system default file
+		 */
+		in = fopen(IANADIR PATH_SEPARATOR IANA_PEN_REGISTRY, "r");
+		if (!in) {
+			lperror(LOG_ERR, "IANA PEN registry open failed");
+			exit(-1); /* Can't leave ipmi_oem_info uninitialized */
```

Comment:
Rather than having the process exit if a PEN list cannot be found, why not just run with an empty one? Sure there will not be complete data, but this seems like a bit of a draconian measure. ipmitool may be wounded without this, but I don't think it is a fatal wound.

Reply:
I agree. Will change it to use an empty dummy array if allocation fails.